File: lib/session.rb
require 'Timeloop'
require_relative 'behavior'
require_relative 'response'

class Session
  attr_reader :duration_in_min, :behaviors, :responses, :start_time, :end_time

  def initialize(duration_in_min = 5)
    @duration_in_min = duration_in_min
    @behaviors = []
    @responses = []
  end

  def add_behavior(new_behavior)
    @behaviors << new_behavior
  end

  def remove_behavior(behavior)
    @behaviors.delete(behavior)
  end

  def duration_in_seconds
    @duration_in_min * 60
  end

  def start
    @start_time = Time.now
    Timeloop.every(1.seconds, maximum: duration_in_seconds) do |i|
      monitor_session(i)
    end
  end

  def end_session
    @end_time = Time.now
    results
  end

  def monitor_session(seconds_elapsed)
    @current_time = seconds_elapsed
    if(seconds_elapsed >= duration_in_seconds)
      end_session
    end
  end

  def track(key, time = 0)
    if(key.downcase == 'q' || key.downcase == 'quit')
      return end_session
    end

    behavior = @behaviors.find { |behavior| behavior.key == key }
    if(behavior.nil?)
      "That behavior does not exist. Try again."
    else
      behavior.increment
      @responses << Response.new(behavior, @current_time)
      "Behavior #{behavior.name} tracked."
    end
  end

  def results
    result_string = "Results:\n"
    result_string << "Session Duration: #{duration_in_min} min\n"

    headers = %w(Behavior Frequency Rate)

    headers.each { |header| result_string << sprintf("%-20s | ", header) }
    result_string << "\n"

    behaviors.each do |behavior|
      result_string << sprintf("%-20s | %-20s | %-20s\n", behavior.name, behavior.frequency, calcuate_rate(behavior))
    end

    result_string
  end

  private

  def calcuate_rate(behavior)
    behavior.frequency / duration_in_min.to_f
  end
end

File: javascript_version/public/javascripts/lib/Session.js
var Session = function(durationInMin){
  this.durationInMin = durationInMin;
  this.behaviors = [];
  this.currentTime = 0;
  this.timerId = null;
  this.callbackTimerId = null;
  this.endTime = 0;
  this.responses = [];
  this.running = false;
  this.completeCallback = null;

  this.durationInSeconds = function(){
    return this.durationInMin * 60;
  };

  this.addBehavior = function(behavior){
    if(behavior){
      this.behaviors.push(behavior);
    }
  };

  this.removeBehavior = function(behavior){
    if(behavior){
      var index = this.behaviors.indexOf(behavior);
      if(index > 0){
        this.behaviors.splice(index, 1);
      }
    }
  };

  this.removeBehaviorByName = function(name){
    if(name){
      for(var i = 0; i < this.behaviors.length; i++){
        if(this.behaviors[i].name === name){
          this.behaviors.splice(i, 1);
        }
      }
    }
  };

  this.track = function(key){
    if(key){
      key = key.toLowerCase();
      if(key == 'q' || key == 'quit'){
        return this.end();
      }

      var selectedBehaviors = this.behaviors.filter(function(behavior){
        return behavior.key == key;
      });

      if(selectedBehaviors.length > 0){
        var selectedBehavior = selectedBehaviors[0];
        selectedBehavior.increment();
        this.responses.push(new Response(selectedBehavior, this.currentTime));
        return selectedBehavior;
      } else {
        return null;
      }
    }
  };

  this.start = function(timerCallback, completeCallback){
    this.startTimer(timerCallback);
    this.running = true;
    if(completeCallback){
      this.completeCallback = completeCallback;
    }
  };

  this.end = function(){
    this.stopTimer();
    this.running = false;
    this.endTime = this.currentTime;
    if(this.completeCallback){
      this.completeCallback();
    }
  };

  this.startTimer = function(callback){
    // need this craziness to maintain context for tick function
    this.timerId = setInterval((
      function(self){
        return function(){ self.tick(); };
      })(this),
      1000);
    if(callback){
      this.callbackTimerId = setInterval(callback, 1000);
    }
  };

  this.stopTimer = function(){
    clearInterval(this.timerId);
    clearInterval(this.callbackTimerId);
  };

  this.tick = function(){
    this.currentTime++;
    if(this.currentTime >= this.durationInSeconds()){
      this.end();
    }
  };

  this.results = function(){
    return new Result(this.endTime, this.behaviors, this.responses);
  };
};
